### ISSUE TITLE
Support for consistent/persistent network device naming

### ISSUE THREAD

**Issue by infernix** [vagrant-libvirt/vagrant-libvirt#609]
In the latest distros (RHEL7, Ubuntu 16.04, others?) NICs are labelled persistently and not simply eth0-x by order of nic detected.

This poses a problem because we rely on the management interface being a DHCP NIC and if a box image isn't sending out dhcp requests on the right nic, we end up nowhere.

We also may have a situation where the a nic ends up named differently because libvirt attaches it to some different bus or slot (e.g. enp2s0f0 or enp1s0f0 and so on) depending on other devices in the VMs XML. This is a tricky case to handle since we need to know beforehand what the NIC is being called in order to do proper detection/configuration.

This can all be worked around by passing `net.ifnames=0 biosdevname=0` on kernel commandline but we should investigate whether if we can fully support persistent naming schemes, provided that box images are configured for that (e.g. by setting DHCP on all possible interfaces).

So if any box image creators/mutators are reading this: use `net.ifnames=0 biosdevname=0` in your grub commandlines until this is fixed.

**Comment by haggaie** [comment]
I have a scenario where I want to add a PCI passthrough NIC that doesn't have a DHCP server on its link. Would it be possible to have some special persistent name to the interface Vagrant/vagrant-libvirt added and enable DHCP on that interface only?

**Comment by infernix** [comment]
@haggaie does the virtio management nic not show up as eth0 and the passthrough one as eth1?

**Comment by haggaie** [comment]
No, the management nic showed up as eth1 and the passthrough as eth0.

**Comment by infernix** [comment]
@haggaie try assigning it a different bus and/or slot number, and/or try picking `e1000` or `rtl8139` for `nic_model_type`.

**Comment by haggaie** [comment]
Is it possible to control the guest bus/slot number via the `libvirt.pci` method? I don't see it in the template. I'll try another nic model type.

**Comment by haggaie** [comment]
Using `e1000` as `nic_model_type` worked. Thanks!

**Comment by ju2wheels** [comment]
Any suggestions on what to look at if we have a Packer built Ubuntu instance for which we can confirm (logging in via virt-manage gui) we have disabled persistent network naming via the following:

```
  sudo sed -ie 's/GRUB_CMDLINE_LINUX="\(.*\)"/GRUB_CMDLINE_LINUX="net.ifnames=0 biosdevname=0 \1"/g' /etc/default/grub
  sudo sed -i 's/auto en.*$/auto eth0/g' /etc/network/interfaces
  sudo sed -i 's/ en.* / eth0 /g' /etc/network/interfaces
  echo 'pre-up sleep 2' | sudo tee -a /etc/network/interfaces
  sudo update-grub
```

But I am still getting no IP on the management interface. Are we required to set a MAC address for the management interface ourselves? I get the following where it cant even find the MAC address that was assigned for the current instance on its own: 

```
==> ansible_test: Creating image (snapshot of base box volume).
==> ansible_test: Creating domain with the following settings...
==> ansible_test:  -- Name:              myproject_ansible_test
==> ansible_test:  -- Domain type:       kvm
==> ansible_test:  -- Cpus:              1
==> ansible_test: 
==> ansible_test:  -- Feature:           acpi
==> ansible_test:  -- Feature:           apic
==> ansible_test:  -- Feature:           pae
==> ansible_test:  -- Memory:            512M
==> ansible_test:  -- Management MAC:    52:54:00:b3:50:8b
==> ansible_test:  -- Loader:            
==> ansible_test:  -- Base box:         myuser/mybox-ubuntu-16.04-ansible-2.5.2
==> ansible_test:  -- Storage pool:      default
==> ansible_test:  -- Image:             /var/lib/libvirt/images/myproject_ansible_test.img (64G)
==> ansible_test:  -- Volume Cache:      default
==> ansible_test:  -- Kernel:            
==> ansible_test:  -- Initrd:            
==> ansible_test:  -- Graphics Type:     vnc
==> ansible_test:  -- Graphics Port:     -1
==> ansible_test:  -- Graphics IP:       127.0.0.1
==> ansible_test:  -- Graphics Password: Not defined
==> ansible_test:  -- Video Type:        cirrus
==> ansible_test:  -- Video VRAM:        9216
==> ansible_test:  -- Sound Type:	
==> ansible_test:  -- Keymap:            en-us
==> ansible_test:  -- TPM Path:          
==> ansible_test:  -- INPUT:             type=mouse, bus=ps2
==> ansible_test: Creating shared folders metadata...
==> ansible_test: Starting domain.
==> ansible_test: Waiting for domain to get an IP address...
```

with a Vagrantfile that is basically:

```
Vagrant.configure('2') do |config|
  config.vm.define 'ansible_test' do |ansible_test|
    ansible_test.vm.box         = 'myuser/mybox-ubuntu-16.04-ansible-2.5.2'
    ansible_test.vm.provider :libvirt do |libvirt, override|
       libvirt.connect_via_ssh   = false
       libvirt.driver            = 'kvm'
       libvirt.storage_pool_name = 'default'
    end
  end
end
```

Im confused as to what needs to be done here to get this instance started with a single NAT interface (which is what I thought the default management interface was).

**Comment by iam-TJ** [comment]
This is becoming even more of an issue now that most distributions are using systemd/udevd > v197 "Predictable Network Interface Names" ( https://www.freedesktop.org/wiki/Software/systemd/PredictableNetworkInterfaceNames/ ).

It is definitely something to be solved in this module rather than requiring what could be disruptive changes to the host configuration via kernel command-line `net.ifnames=0 biosdevname=0`.

I've just started working with vagrant + vagrant-libvirt on Ubuntu 18.04 and lost a day trying to diagnose why a basic ubuntu/bionic64 box fails to start with:

`Call to virDomainCreateWithFlags failed: Unable to get index for interface eth0: No such device"`

which is directly from libvirt (same error in syslog).

For many hours I thought it was due to my not getting a `config.vm.network :public_network, ...` statement correct. Finally I commented it out and the error still occurred!

At that point I realised the culprit is the 'hidden' `management_network` configuration. I dived into the source-code and see that there is no attribute exposed to set the `@device` value and therefore it uses the default `eth0`:

`lib/vagrant-libvirt/action/create_network_interfaces.rb:85:              @device = iface_configuration.fetch(:dev, 'eth0')`

This needs documenting as a gotcha up-front for newcomers especially (in docs and in the generated Vagrantfile) and an attribute provided to over-ride the interface device name at a minimum.

It would be preferable to detect the host's interface. Manually that is achieved with:

`brctl show  $(virsh net-info vagrant-libvirt | awk '/^Bridge:/{print $2}' ) | awk 'NR != 1{print $4}'`

I did try working around it using an alias, but that isn't exposed in any useful way as an alternate name, as in:

`sudo ip link set virbr1-nic alias eth0`

**Comment by iam-TJ** [comment]
I've just noticed after pulling the source that an attribute has been added as of:

commit 7f50ca2a41bbaab573132ba00ec552f28a93c3cb
Author: qazokm <qazokm@users.noreply.github.com>
Date:   Fri Feb 23 12:09:49 2018 +0800

    Add basic networking support for qemu session

which has added:

`
@management_network_device = 'virbr0' if @management_network_device == UNSET_VALUE
@system_uri      = 'qemu:///system' if @system_uri == UNSET_VALUE
`
